multi/imxrt-multi/spi.c

```diff
@@ -0,0 +1,441 @@
+
+	/* Wait until Transmit Command will be taken from  TX fifo */
+	txWordsCnt = *(spi_common[spi].base + spi_fsr) & 0x1f;
+	while (txWordsCnt)
+		txWordsCnt = ((*(spi_common[spi].base + spi_fsr)) & 0x1f);
+
+	while (size > 0) {
+		if (size <= MAX_FIFOSZ_BYTES)
+			txFifoBytes = size;
+		else
+			txFifoBytes = MAX_FIFOSZ_BYTES;
+
+		/* Fill transmit FIFO */
+		spi_txBytes(spi, txBuff, txFifoBytes);
+		txBuff += txFifoBytes;
+		size -= txFifoBytes;
+
+		/* Transfer data into slave */
+		txWordsCnt = *(spi_common[spi].base + spi_fsr) & 0x1f;
+		spi_initTransmition(spi);
+
+		/* Wait on transmit data.
+		 * Omit this step if Transmit Command is sent in transfers bigger than MAX_FIFOSZ_BYTES */
+		rxWordsCnt = (*(spi_common[spi].base + spi_fsr) >> 16) & 0x1f;
+		while ((rxWordsCnt != txWordsCnt) & (len <= MAX_FIFOSZ_BYTES))
+			rxWordsCnt = (*(spi_common[spi].base + spi_fsr) >> 16) & 0x1f;
+
+		/* Get data from receive FIFO */
+		rxFifoBytes = rxWordsCnt * WORD_SIZE;
+		if ((rxFifoBytes + rxTotalBytes) > len)
+			rxFifoBytes = len - rxTotalBytes;
+
+		spi_rxBytes(spi, rxBuff, rxFifoBytes);
+		rxBuff += rxFifoBytes;
+		rxTotalBytes += rxFifoBytes;
+	}
+
+	/*
+	 * If data length is a multiplication of MAX_FIFOSZ_BYTES, Transmit Command moves one word out of FIFO.
+	 * The last word is received after trasnsmission.
+	 */
+	rxFifoBytes = len - rxTotalBytes;
+
+	if (rxFifoBytes) {
+		rxWordsCnt = (*(spi_common[spi].base + spi_fsr) >> 16) & 0x1f;
+		if (rxWordsCnt) {
+			spi_rxBytes(spi, rxBuff, rxFifoBytes);
+			rxTotalBytes += rxFifoBytes;
+		}
+	}
+
+	return rxTotalBytes;
+}
+
+
+static int spi_configure(uint32_t spi, uint32_t bdiv, uint32_t prescaler, uint32_t endian, uint32_t mode, uint32_t cs)
+{
+	int i;
+
+	if (!spiConfig[spi])
+		return -EINVAL;
+
+	if (bdiv > 255)
+		return -EINVAL;
+
+	if (prescaler >= 8)
+		return -EINVAL;
+
+	i = spiPos[spi];
+
+	/* Disable module */
+	*(spi_common[i].base + spi_cr) = 0;
+
+	/* Software reset */
+	*(spi_common[i].base + spi_cr) |= 1 << 1;
+	*(spi_common[i].base + spi_cr) = 0;
+
+	/* Set master mode as default */
+	*(spi_common[i].base + spi_cfgr1) |= 1;
+
+	/* CS polarity: 0 - low */
+	*(spi_common[i].base + spi_cfgr1) &= ~(0xf << 8);
+
+	/* Set pin configuration */
+	*(spi_common[i].base + spi_cfgr1) &= ~(1 << 26) & ~(3 << 24);
+
+	/* Enable Module */
+	*(spi_common[i].base + spi_cr) |= 1;
+
+	/* Set baudrate divider */
+	*(spi_common[i].base + spi_ccr) = (*(spi_common[i].base + spi_ccr) & ~0xff) | bdiv;
+
+	/* Set watermarks */
+	*(spi_common[i].base + spi_fcr) &=  ~(0xf << 16) & ~0xf;
+
+	/* Set transmit command register */
+	*(spi_common[i].base + spi_tcr) &= ~(1 << 26); /* reset register */
+	spi_common[i].tcr = *(spi_common[i].base + spi_tcr) | ((mode & 0x3) << 30) | ((cs & 0x3) << 24) | ((endian & 0x1) << 23) | (prescaler << 27);
+
+	interrupt(spi_common[i].irq, spi_irqHandler, (void *)i, spi_common[i].cond, &spi_common[i].inth);
```

Comment:
You **do not** want to set an interrupt at each configuration event. You should set interrupts handlers for every supported spi in initialization routine, only once